### PR TITLE
planner: fix the issue that some PointGet plans generated in physical-stage cannot be cached

### LINKED ISSUE #26868: prepared statement doesn't work on 'insert into t1 select  from t2 where id = ?'
## Enhancement

The query is 'insert into t1 select  from t2 where id = ?', it inserts a single line each time, repeat 1000 times with 200 concurrency.

The plan cache doesn't take effect, and it seems to be very inefficient.
Even worse, I suspect in this case it's even slower than the non-prepared statement.
As you can see, most of the time are spend on rebuild of the plan,  

image

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: close #26868 

Problem Summary: planner: fix the issue that some PointGet plans generated in physical-stage cannot be cached

### What is changed and how it works?

planner: fix the issue that some PointGet plans generated in physical-stage cannot be cached

### Check List 

Tests 

- [x] Unit test

### Release note 

```release-note
planner: fix the issue that some PointGet plans generated in physical-stage cannot be cached
```

### CODE REVIEW

File: expression/expression.go [hunk 1]
```diff
@@ -789,8 +789,8 @@ func SplitDNFItems(onExpr Expression) []Expression {
 // EvaluateExprWithNull sets columns in schema as null and calculate the final result of the scalar function.
 // If the Expression is a non-constant value, it means the result is unknown.
 func EvaluateExprWithNull(ctx sessionctx.Context, schema *Schema, expr Expression) Expression {
-	if ContainMutableConst(ctx, []Expression{expr}) {
-		ctx.GetSessionVars().StmtCtx.OptimDependOnMutableConst = true
+	if ctx.GetSessionVars().StmtCtx.UseCache && ContainMutableConst(ctx, []Expression{expr}) {
```
Comment (Reminiscent) [inline expression/expression.go:792]: The function `ContainMutableConst` contains the check for `ctx.GetSessionVars().StmtCtx.UseCache`.

File: planner/core/prepare_test.go [hunk 2]
```diff
@@ -1107,7 +1149,7 @@ func (s *testPlanSerialSuite) TestPlanCachePointGetAndTableDual(c *C) {
 	tk.MustQuery("select @@last_plan_from_cache").Check(testkit.Rows("0"))
 	// Must not reuse the previous TableDual plan.
```
Comment (Reminiscent) [inline planner/core/prepare_test.go:1150]: Please modify the comment here. Besides, we need to check the plan from the cache.

File: planner/core/stats.go [hunk 3]
```diff
@@ -604,7 +602,7 @@ func (ds *DataSource) isInIndexMergeHints(name string) bool {
 }
 
 // accessPathsForConds generates all possible index paths for conditions.
```
Comment (Reminiscent) [inline planner/core/stats.go:604]: Please update the comment here.

File: planner/core/stats.go [hunk 4]
```diff
@@ -343,8 +343,6 @@ func (ds *DataSource) derivePathStatsAndTryHeuristics() error {
 	if selected != nil {
 		ds.possibleAccessPaths[0] = selected
 		ds.possibleAccessPaths = ds.possibleAccessPaths[:1]
-		// TODO: Can we make a more careful check on whether the optimization depends on mutable constants?
-		ds.ctx.GetSessionVars().StmtCtx.OptimDependOnMutableConst = true
```
Comment (Reminiscent) [inline planner/core/stats.go:347]: Are there some test cases that can cover this change?

File: planner/core/find_best_task.go [hunk 5]
```diff
@@ -811,7 +811,10 @@ func (ds *DataSource) findBestTask(prop *property.PhysicalProperty, planCounter
 				p: dual,
 			}, cntPlan, nil
 		}
-		canConvertPointGet := len(path.Ranges) > 0 && path.StoreType == kv.TiKV && ds.isPointGetConvertableSchema()
+		canConvertPointGet := len(path.Ranges) > 0 && path.StoreType == kv.TiKV && ds.isPointGetConvertableSchema() &&
+			// to avoid the over-optimized risk, do not generate PointGet for plan cache, for example,
+			// `pk>=$a and pk<=$b` can be optimized to a PointGet when `$a==$b`, but it can cause wrong results when `$a!=$b`.
+			!ds.ctx.GetSessionVars().StmtCtx.UseCache
```
Comment (xuyifangreeneyes) [inline planner/core/find_best_task.go:817]: Here we cannot generate `PointGet` if prepared-cache-plan is enabled. Will that be a problem in OLTP workload? For example, as for `prepare stmt1 from "insert into t2 select * from t1 where a=?"`, if the sql will be executed frequently, we may hope to not only generate `PointGet` plan but also cache the plan.

File: util/ranger/detacher.go [hunk 6]
```diff
@@ -566,7 +566,9 @@ func ExtractEqAndInCondition(sctx sessionctx.Context, conditions []expression.Ex
 				// Maybe we can improve it later.
 				columnValues[i] = &valueInfo{mutable: true}
 			}
-			sctx.GetSessionVars().StmtCtx.OptimDependOnMutableConst = true
+			if sctx.GetSessionVars().StmtCtx.UseCache {
```
[the remark this answers is not in the document]
Comment (xuyifangreeneyes) [reply util/ranger/detacher.go:569]: In real workloads there may be a small chance to convert `a > ? and  a < ?` to equal condition, so maybe it is acceptable to give up the optimization when prepare-plan-cache is enabled.

**Comment by qw4990** [comment]
There are too many conflicts, I'll close this PR and submit a new one later.

**Pull request closed without merging**